src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -78,29 +83,49 @@ struct TensorContract
     : public TensorExpression<TensorContract<ReplacedArg1, ReplacedArg2, T, X,
                                              Symm, IndexList, ArgsList>,
                               X,
-                              typename detail::ComputeContractedType<
+                              typename detail::ContractedType<
                                   ReplacedArg1, ReplacedArg2, T, X, Symm,
-                                  IndexList, ArgsList>::symmetry,
-                              typename detail::ComputeContractedType<
+                                  IndexList, ArgsList>::type::symmetry,
+                              typename detail::ContractedType<
                                   ReplacedArg1, ReplacedArg2, T, X, Symm,
-                                  IndexList, ArgsList>::index_list,
-                              typename detail::ComputeContractedType<
+                                  IndexList, ArgsList>::type::index_list,
+                              typename detail::ContractedType<
                                   ReplacedArg1, ReplacedArg2, T, X, Symm,
-                                  IndexList, ArgsList>::args_list> {
-  static constexpr size_t Index1 =
-      tmpl::index_of<ArgsList, TensorIndex<ReplacedArg1::value>>::value;
-  static constexpr size_t Index2 =
-      tmpl::index_of<ArgsList, TensorIndex<ReplacedArg2::value>>::value;
-  using CI1 = tmpl::at_c<IndexList, Index1>;
-  using CI2 = tmpl::at_c<IndexList, Index2>;
+                                  IndexList, ArgsList>::type::args_list> {
+  static constexpr size_t replaced_arg1_position =
+      tmpl::index_of<ArgsList, ReplacedArg1>::value;
+  static constexpr size_t replaced_arg2_position =
+      tmpl::index_of<ArgsList, ReplacedArg2>::value;
+
+  // "first" and "second" here refer to the position of the indices to contract
+  // in the list of generic indices, with "first" denoting leftmost
+  //
+  // e.g. `R(ti_A, ti_b, ti_a)` :
+  // - `first_replaced_tensorindex` refers to the contracted TensorIndex type
+  //   for `ti_A`
+  // - `second_replaced_tensorindex` refers to the contracted TensorIndex type
+  //   for `ti_a`
+  static constexpr size_t first_position_to_contract =
+      (replaced_arg1_position < replaced_arg2_position)
+          ? replaced_arg1_position
+          : replaced_arg2_position;
+  static constexpr size_t second_position_to_contract =
+      (replaced_arg1_position < replaced_arg2_position)
+          ? replaced_arg2_position
+          : replaced_arg1_position;
+
+  // first and second \ref SpacetimeIndex "TensorIndexType"s to contract
+  using CI1 = tmpl::at_c<IndexList, first_position_to_contract>;
```

Comment:
what about naming these `contracted_upper_index` and `contracted_lower_index`?

Reply:
I was on the fence about doing this here in this PR because there are a number of aliases I want to rename, but I'll do this one now!

Reply:
Done